bitcoin-cli: Add -ipcconnect option

ryanofsky opened the pull request:
This implements an idea from sipa in https://github.com/bitcoin/bitcoin/issues/28722#issuecomment-2807026958 to allow `bitcoin-cli` to connect to the node via IPC instead of TCP, if the ENABLE_IPC cmake option is enabled and the node has been started with `-ipcbind`.

This feature can be tested with:

```
build/bin/bitcoin-node -regtest -ipcbind=unix -debug=ipc
build/bin/bitcoin-cli -regtest -ipcconnect=unix -getinfo
```

The -ipconnect parameter can also be omitted, since this change also makes `bitcoin-cli` prefer IPC over HTTP by default, and falling back to HTTP if an IPC connection can't be established.

---

This PR is part of the [process separation project](https://github.com/bitcoin/bitcoin/issues/28722).

ryanofsky committed:
ipc: Improve -ipcconnect error checking

When an invalid socket path is passed to -ipcconnect, either because the path
exceeds the maximum socket length, or the path includes a directory component
which is not actually a directory, treat this the same as the same as the
socket refusing connections or not existing, instead of treating it like a more
serious I/O error and throwing a fatal exception.

This is needed to avoid CI errors after the following commit which adds a
functional test and uses -datadir paths exceeding the maximum socket length
when running in CI.

ryanofsky committed:
refactor: Add ExecuteHTTPRPC function

Add ExecuteHTTPRPC to provide a way to execute an HTTP request without relying
on HTTPRequest and libevent types.

Behavior is not changing in any way, this is just moving code. This commit may
be easiest to review using git's --color-moved option.

ryanofsky committed:
ipc: Add MakeBasicInit function

Add a MakeBasicInit() function so simpler standalone IPC clients like
bitcoin-mine in #30437 and bitcoin-cli in #32297 that only initiate IPC
connections without exposing any IPC interfaces themselves can to avoid needing
to implement their own specialized interfaces::Init subclasses.

ryanofsky committed:
bitcoin-cli: Add -ipcconnect option

This implements an idea from Pieter Wuille <pieter@wuille.net>
https://github.com/bitcoin/bitcoin/issues/28722#issuecomment-2807026958 to
allow `bitcoin-cli` to connect to the node via IPC instead of TCP, if the
`ENABLE_IPC` cmake option is enabled and the node has been started with
`-ipcbind`.

The feature can be tested with:

build/bin/bitcoin-node -regtest -ipcbind=unix -debug=ipc
build/bin/bitcoin-cli -regtest -ipcconnect=unix -getinfo

The `-ipconnect` parameter can also be omitted, since this change also makes
`bitcoin-cli` prefer IPC over HTTP by default, and falling back to HTTP if an
IPC connection can't be established.

laanwj commented on src/bitcoin-cli.cpp:
```diff
@@ -943,9 +964,26 @@ static UniValue ConnectAndCallRPC(BaseRequestHandler* rh, const std::string& str
+        }
+    }
+
+    std::string username{gArgs.GetArg("-rpcuser", "")};
```
Passing `-rpcconnect` and `-ipcconnect` at the same time should probably result in an error.

laanwj commented:
Concept ACK, it'd be useful to have an utility that can talk to the node through IPC and making `bitcoin-cli` double for that prevents code duplication.

jlest01 commented:
Concept ACK, as it can make communication between the CLI and the node simpler and configuration-free.

pinheadmz commented:
concept ACK, cool idea and will review. Does this close https://github.com/bitcoin/bitcoin/issues/5029 (adding unix sockets to the bitcoind rpc api) or does it only apply to `bitcoin-node`? Or is IPC protocol different from RPC and must be defined in the client?

ryanofsky commented:
> concept ACK, cool idea and will review. Does this close #5029 (adding unix sockets to the bitcoind rpc api) or does it only apply to `bitcoin-node`? Or is IPC protocol different from RPC and must be defined in the client?

It uses capnproto, so it may or may not achieve goals of #5029 depending on what they are. This PR does make it possible to call RPC methods over a unix socket, but in order to do that you need to use capnproto which is not available everywhere. For example there wouldn't be an easy way for our python test framework to use the socket directly unless it depended on https://github.com/capnproto/pycapnp. So another solution that uses a another protocol or just uses http over the socket might be preferable to this approach. But the first commit here could be useful for supporting other approaches as well.

ryanofsky committed:
test: add interface_ipc_cli.py testing bitcoin-cli -ipcconnect

Co-authored-by: Sjors Provoost <sjors@sprovoost.nl>

ryanofsky reviewed (commented):
Updated df40f46571bc7b8f0a85a437130e99bad5b0de63 -> 538521a37c3788ce2a0ba1bda76844a25cfd3e06 ([`pr/ipc-cli.2`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-cli.2) -> [`pr/ipc-cli.3`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-cli.3), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-cli.2..pr/ipc-cli.3)) handling -rpcconnect option, adding a test, making various cleanups, adding comments, and fixing CI failures in non-ipc builds
Updated 538521a37c3788ce2a0ba1bda76844a25cfd3e06 -> bd29de3896911317fc6dc0010c33b91ebf5b9f16 ([`pr/ipc-cli.3`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-cli.3) -> [`pr/ipc-cli.4`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-cli.4), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-cli.3..pr/ipc-cli.4)) fixing compile error in intermediate commit, https://github.com/bitcoin/bitcoin/actions/runs/14538988101/job/40793056407?pr=32297, fixing functional test failure when socket path is too long https://cirrus-ci.com/task/6103052337283072, and adding new test to cover "bitcoin-cli was not built with IPC support" error

yancyribbens commented:
Concept ACK

luke-jr commented:
This looks like it uses JSON-RPC inside Capnproto? Why not just use straight Capnproto?

ryanofsky commented:
> This looks like it uses JSON-RPC inside Capnproto? Why not just use straight Capnproto?

Yes it is skipping the HTTP part of the protocol but keeping the JSON part. The capnproto interface this exposes looks like:

```
interface Rpc $Proxy.wrap("interfaces::Rpc") {
    executeRpc @0 (context :Proxy.Context, request :Text, uri :Text, user :Text) -> (result :Text);
}
```

where `request` and `response` above are JSON text strings with the expected JSON-RPC fields. The schema could be unwrapped further, or the data could be sent in some binary json format depending on what your goal is but that interface was the simplest way I could think of implementing this feature.

shahsb reviewed (commented):
Concept ACK

laanwj commented:
> The schema could be unwrapped further, or the data could be sent in some binary json format depending on what your goal is but that interface was the simplest way I could think of implementing this feature.

Right, Say, representing the entire RPC protocol as cap'n'proto structures directly would be interesting (anologous to how c-lightning has a gRPC version of their RPC), but this would be a huge project for which the first step would be the formal definition of the RPC protocol (including specific-enough types) so that that can be automatically generated. 

And this would make it harder to make a cli tool, not easier, because the client needs to know the protocol to convert its command line to suitable calls.

So sending the JSON inside the interface seems like a good, low-impact way to implement this tool.

pinheadmz commented on src/httprpc.h:
```diff
@@ -19,6 +23,9 @@ void InterruptHTTPRPC();
  */
 void StopHTTPRPC();
 
+/** Execute a single HTTP request, containing one or more JSONRPC requests. */
```
ea98a42640b9ff77a462e55887025ddd1da54727

Might want to indicate `status` is an "out" param?

I notice this also bypasses our only authorization barrier for RPC which is an HTTP header, maybe you deal with this in another commit but that should probably be documented in the `ipcbind` help text

pinheadmz commented:
While reviewing this I hit some weird behavior where `bitcoin-node` doesn't fully shutdown.

To reproduce:

In terminal 1: `build/bin/bitcoin-node -regtest -ipcbind=unix -debug=rpc -debug=http -debug=ipc`

Terminal 2: `build/bin/bitcoin-cli -regtest waitforblock 0000000000079f8ef3d2c688c244eb7a4570b24c9ed7b4a8c619eb02596f8862`
(this will wait forever)

go back to terminal 1 and ctrl-c. I get `Shutdown: done` in the output but the process does not quit.

Then funny stuff like this, in Terminal 3:

```
--> build/bin/bitcoin-cli -regtest getblockcount
error code: -32603
error message:
Node chainman not found

--> build/bin/bitcoin-cli -regtest -getinfo
error code: -1
error message:
Internal bug detected: RPC call "getnetworkinfo" returned incorrect type:
{
    "localservices": "key missing, despite not being optional in doc",
    "localservicesnames": "key missing, despite not being optional in doc",
    "localrelay": "key missing, despite not being optional in doc",
    "timeoffset": "key missing, despite not being optional in doc",
    "connections": "key missing, despite not being optional in doc",
    "connections_in": "key missing, despite not being optional in doc",
    "connections_out": "key missing, despite not being optional in doc",
    "networkactive": "key missing, despite not being optional in doc",
    "relayfee": "key missing, despite not being optional in doc",
    "incrementalfee": "key missing, despite not being optional in doc"
}
Bitcoin Core v29.99.0-9ff5bc75659e
Please report this issue here: https://github.com/bitcoin/bitcoin/issues

```

I had to use `killall -9 bitcoin-node` to clean up

ryanofsky commented:
> While reviewing this I hit some weird behavior where `bitcoin-node` doesn't fully shutdown.

Thanks for testing! These are all known issues which should be fixed by https://github.com/bitcoin/bitcoin/pull/32345

pinheadmz reviewed (approved):
ACK 9ff5bc75659e0314f42c1477fc0c37ac0ad132fa

Built and ran tests on arm64/macos. Played with the feature locally. Reviewed each commit, some questions below.

<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK 9ff5bc75659e0314f42c1477fc0c37ac0ad132fa
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmgT0ykACgkQ5+KYS2KJ
yTogNQ/+NxdemF18NTDuoW6W7fLeDbdbJCbe/9fSse0fhFi0rtTkFpJdXKt+p86T
gndlLfTNitJ58BUul5WMbJvotwnydkBoSlsPi1ZT1hWainWuZ7Q6fkYgRV94uuU5
arsNYxwOHOlzwu/mbcxvkUgDSWHqLS0zaw1kNfPby6zcF0AajUS/gtAwNObv7h9C
zMfLjm537So8fbiimSxnF1N4aHchhd63AVm8PUD+X9E1ZRN0xc7IeWfGcUc0W2Zx
kAxwOKO1kQBBDqHlktlQ8F+cSUjdNXI4JXqb6mCNPXH4WFk669xsBV3Yz8Z/73gt
lCBk7spyO/TWMdhdZU4NrGRvkgBbn5ZgjRWDIkHP/9JZYfunoBASbsGI1Oiexolz
t2CTZGDKvdkKD3Zz1aYRKMbeF4uBxY0HA4sPauwt6N9tbtk3Mj2nUBbLQgMO4the
4Lx6IT0ewpBcfbQq+8/7rbuDFvAh+v11oqwUbvckXYuZxkv6epU/3kjyoZOmJ3Xd
afs6rt3BHRk6HnETLKq8fZgNc6tkmz7rrRPZSYgHw039BStK/MG9SowGxIFtu8Wj
iYrfZcSVeIeFkag67HDpHnLAaDV3jneZuvocEynypk8R4LHd9TGH+b4KEFj6No1d
F8Nhx3FyY6AeuGlWKz3lSxEZlqKB/eC4chyxzTWwkHVfIDA8KGU=
=IN//
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on openpgp.org](https://keys.openpgp.org/vks/v1/by-fingerprint/E61773CD6E01040E2F1BD78CE7E2984B6289C93A)
</details>

ryanofsky reviewed (commented):
Great review! I improved a number of things in response to your comments.

Rebased 9ff5bc75659e0314f42c1477fc0c37ac0ad132fa -> 6bf727fb88009f99dab7c06bff7f8fb391147ac1 ([`pr/ipc-cli.5`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-cli.5) -> [`pr/ipc-cli.6`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-cli.6), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-cli.5-rebase..pr/ipc-cli.6)) due to conflict with #28710, also made a number of improvements to -ipcbind documentation and -rpcwait behavior and other code as suggested, also split commits and cleaned up bitcoin-cli build rule and new tests.
Rebased 6bf727fb88009f99dab7c06bff7f8fb391147ac1 -> ad0144022c2bbed6799bafa3ac8b80ac81dd7178 ([`pr/ipc-cli.6`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-cli.6) -> [`pr/ipc-cli.7`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-cli.7), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-cli.6-rebase..pr/ipc-cli.7)) due to conflict with #31375

pinheadmz reviewed (approved):
re-ACK ad0144022c2bbed6799bafa3ac8b80ac81dd7178

reviewed each commit again, built and ran tests on macos/arm64. Played with the feature on main, requesting large block data and `waitforblockheight` etc. Multiple requests can be executed in parallel over the same socket which is very cool, works with bitcoin-node and bitcoin-gui, tested both. Although `waitforblockheight` doesn't work with bitcoin-gui because `EnsureMiner` fails. That was a surprise to me but may just be expected.

Two non blocking nits


<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK ad0144022c2bbed6799bafa3ac8b80ac81dd7178
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmhRrOEACgkQ5+KYS2KJ
yTqgLhAA1fi5wv1y2LjeIPUqmzllmoBYnmlfI+Pki0cOFIH60DMhYLYrBc7gW61x
1U9G39GbscmOF6OO+z9viST62JlXTdhDhs3CieCaRX8n2QNGwnKRh8wO6jugybol
55XAvj5xJ7DwSzYJrQH8yMrzL/atHwkgC9ut0U4z2F3JgI0Pbt+PKhzxF7vwQFXB
BNH6xy0guuj2FEZ+jnicugpcABqKVtaypFZREs7EuESVfSCCTVbiuM2DLkbp1OKm
5aRnyifd9p0JNn71ApoJEcpx6shjWSo4fsVS26FmnW/+vyYKJ3y4wUctcLJSitVx
YtfeJ0wxlAa9PtkcHUkw/C0+DmW3aI1mK9N2rxXGIqOQ1fQVk270dp7nvpdcQbds
adhTRd/sm0kt9f83MwVwqeFWpuVoLEAQa2T6HTaVP9lSDvT+I3abZmR1dOwu8EG+
r0U0MrL5gMgOhLIbUXTBt9Heyo725ODaPiYJin/gmytAbL+WYJWG7hIardNg9vY/
74OtFw/Dd12DixO1IamkV0SiH7PqHpxnOZgJ+D+YLF0JilPXEpNOZ9CVFmX5++/D
VwOq+fEwo7l3c/CVt/7NRQrJTf8MeYkA2epVlBmUl8/ITihJ2qigQiqIBn91Pr5S
qt7HPHww2eRTYdqT+HHnTIkkFDywM5nToA5THA5Y+pyMgScVQ0U=
=UEB3
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on openpgp.org](https://keys.openpgp.org/vks/v1/by-fingerprint/E61773CD6E01040E2F1BD78CE7E2984B6289C93A)
</details>

ryanofsky committed:
bitcoin-gui: Implement missing Init::makeMining method

A missing Init::makeMining implementation was causing internal code using the
mining interface (like the `waitforblockheight` RPC method) to not work when
running inside the `bitcoin-gui` binary. It was working the other bitcoin
binaries ('bitocind`, `bitcoin-qt`, and `bitcoin-node`) because they
implmemented `Init::makeMining` methods in commit
8ecb6816781c7c7f423b501cbb2de3abd7250119 from #30200, but the `bitcoin-gui`
init class was forgotten in that change.

This bug was reported by Matthew Zipkin <pinheadmz@gmail.com>
https://github.com/bitcoin/bitcoin/pull/32297#pullrequestreview-2932651216

ryanofsky reviewed (commented):
Thanks for the review and testing!

Updated ad0144022c2bbed6799bafa3ac8b80ac81dd7178 -> 9ffe57f81b2a6abd161ae010f07916b05e57191d ([`pr/ipc-cli.7`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-cli.7) -> [`pr/ipc-cli.8`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-cli.8), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-cli.7..pr/ipc-cli.8)) with suggested `bitcoin-node` fix and documentation update.

---

re: https://github.com/bitcoin/bitcoin/pull/32297#pullrequestreview-2932651216

> Although waitforblockheight doesn't work with bitcoin-gui because EnsureMiner fails. That was a surprise to me but may just be expected.

Good catch. This is bug caused by not defining a `makeMining` method in 8ecb6816781c7c7f423b501cbb2de3abd7250119 from #30200 which should be fixed by 9ffe57f81b2a6abd161ae010f07916b05e57191d in the latest push. Also I think the bug not specific to this PR and happens in master too with RPC.

janb84 reviewed (commented):
ACK 9ffe57f81b2a6abd161ae010f07916b05e57191d

PR adds functionality to allow bitcoin-cli to connect to the node via IPC instead of TCP (if enabled) as part of the Multiprocess project

nit maybe fix grammar issues found by drahtbot https://github.com/bitcoin/bitcoin/pull/32297#issuecomment-2813015247 ? 

- code review ✅
- build & tested ✅
- tested communication between cli and node binaries ✅

ryanofsky reviewed (commented):
<!-- begin push-9 -->
Rebased 9ffe57f81b2a6abd161ae010f07916b05e57191d -> 37cd2c076434e7acbdbb20996cf87afb2cb5bc84 ([`pr/ipc-cli.8`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-cli.8) -> [`pr/ipc-cli.9`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-cli.9), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-cli.8-rebase..pr/ipc-cli.9))<!-- end --> to resolve conflicts and also updated with suggestions.
<!-- begin push-10 -->
Rebased 37cd2c076434e7acbdbb20996cf87afb2cb5bc84 -> c800ccbe8872ec6887587f65976994884981812e ([`pr/ipc-cli.9`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-cli.9) -> [`pr/ipc-cli.10`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-cli.10), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-cli.9-rebase..pr/ipc-cli.10))<!-- end -->
<!-- begin push-11 -->
Updated c800ccbe8872ec6887587f65976994884981812e -> c5e446362c28fa0d296fe6580abccf74d808a7e3 ([`pr/ipc-cli.10`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-cli.10) -> [`pr/ipc-cli.11`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-cli.11), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-cli.10..pr/ipc-cli.11))<!-- end --> to fix silent conflict with #33201
<!-- begin push-12 -->
Rebased c5e446362c28fa0d296fe6580abccf74d808a7e3 -> 65a43932b61582179ebc3ec851172e4b257648e1 ([`pr/ipc-cli.11`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-cli.11) -> [`pr/ipc-cli.12`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-cli.12), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-cli.11-rebase..pr/ipc-cli.12))<!-- end --> due to conflicts with #33960 and #34006
<!-- begin push-13 -->
Rebased 65a43932b61582179ebc3ec851172e4b257648e1 -> b23e42a8ea1d1db649aaa25c1c286b17ec7dcf39 ([`pr/ipc-cli.12`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-cli.12) -> [`pr/ipc-cli.13`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-cli.13), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-cli.12-rebase..pr/ipc-cli.13))<!-- end --> due to conflict with #33192
<!-- begin push-14 -->
Rebased b23e42a8ea1d1db649aaa25c1c286b17ec7dcf39 -> 4565cff72c60f56d8ad4b46716fe8df847554824 ([`pr/ipc-cli.13`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-cli.13) -> [`pr/ipc-cli.14`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-cli.14), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-cli.13-rebase..pr/ipc-cli.14))<!-- end --> due to conflict with #34661

janb84 reviewed (commented):
re ACK 37cd2c076434e7acbdbb20996cf87afb2cb5bc84

changes since last ACK:
- rebase
- spelling/grammer fix
- changes due to suggestions (thanks !) 

build & tested ✅

BrandonOdiwuor reviewed (commented):
Tested ACK 37cd2c076434e7acbdbb20996cf87afb2cb5bc84

Tested on macOS 15.5

Was able to connect `bitcoin-cli` to `bitcoin-node`  successfully over `ipc`

<img width="837" height="453" alt="Screenshot 2025-07-12 at 11 24 10" src="https://github.com/user-attachments/assets/179360fd-5572-44c0-b581-6afe50e7223c" />

- Also confirmed that `bitcoin-cli` uses `IPC` as the preferred mode of connection over `HTTP` even when the `-ipcconnect` option is not provided

pinheadmz reviewed (approved):
re-ACK 37cd2c076434e7acbdbb20996cf87afb2cb5bc84

Changes since last ACK:
`git range-diff ad01440~6..ad01440 37cd2c0764~9..37cd2c0764`
Address PR feedback adding a line to interface docs, add mining interface to the GUI for testing, and refactors the test commits.

I rebased this PR at this commit on to #32345 and pummeled the IPC-RPC interface for an hour with no crash, so I think it's important that those libmutliprocess changes get merged as well, and *probably* should be merged first?

<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK 37cd2c076434e7acbdbb20996cf87afb2cb5bc84
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmiUsx0ACgkQ5+KYS2KJ
yToNXg//byf9IRhhpTQ39dvrI4qC9R7ZLo8fCp9yQPP+7Is6mwSVlfVtfRfXlRqa
EGCxTxQttN2hcWyyqdt2+gIQDP/ytROECGWTs7iR24jJcXE01JmM0N42zNKfi949
kUIwVDtLJcN45gEKjVf1DUH/kUAB63VQLAimqcop15uqc0ZFYynxEktyox2/yAEq
MWQ04ctpF3NR2LccIY0pdR7n3zgKe7HWmAVQgh8/AdlxawM0wKMmFheQXF/7jVhM
yYKKkY1Zst/02GmQw3Kfe2mLOmFol0L6EuGuCrQzTSRV9/mlV1Ba8GiGGbjlfwWp
KPQ9ILwTQfCn0LqQ1o2DWuqvCnYIKjajbDcaTd5VO0p1pnQRtYo8NfYVgXz99snG
AeVRyBQO9gvj39nI4PgH9kLK3uW+bxIGwEHRPuAYrhkOGPStUgZa9EAB9M1C6Sry
KVs9l89KZL+XgC1u6DMe8WQUXnBBNtDvp53czw7n6GJv0R4C6pZSiI6qZ5VJ4ouq
67F3vkaKN2GA6mqH2JYYpDL26UiraSW/cg/sXRaGPxLhp/SDzxRSER4/zOxB0uZr
aRNoMRZM9LGkMEkVuZSmBejsAZyxzIQhiJSToqnw0JFqnc7jV524+NoW2r11b1yv
+YNhVjVhXQSgdZ1/XIsAvuhwCYIJdIAuoHw8sJwJzLtFJ4Z2RhI=
=Yr0Y
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on openpgp.org](https://keys.openpgp.org/vks/v1/by-fingerprint/E61773CD6E01040E2F1BD78CE7E2984B6289C93A)
</details>

ryanofsky committed:
ipc: Expose an RPC interface over the -ipcbind socket

This allows `bitcoin-cli` to connect to the node via IPC instead TCP to execute
RPC methods in an upcoming commit.

sipa reviewed (commented):
Concept ACK and mild code review ACK.

ryanofsky commented:
Yep, it's ready for review. I don't think there have been any real changes. The commits that were dropped just became part of #33201, IIRC

pinheadmz reviewed (approved):
ACK b23e42a8ea1d1db649aaa25c1c286b17ec7dcf39

Built and tested on arm64/macos. Re-reviewed each commit to refresh my memory on how interfaces work. A few questions below, nothing breaking but I am curious about the choice of `ExecuteHTTPRPC` vs `JSONRPCExec()`.

<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK b23e42a8ea1d1db649aaa25c1c286b17ec7dcf39
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmmV3sMACgkQ5+KYS2KJ
yToxDhAAi75OL2MuiaURUSlGdf3zDKrodG4FRByGnFEas6ztTB/RpFCNd4cc+LUm
wGcPRXTLlYcVb1NwZAgrL20C6jy28Sy6NW/d4saKRHHO0ugqFmkL8OgxHaGEI7oL
nvZrw+fyZD9MBjtzy4unVURLKVBdQP5+KySC9E7b2JJmvNU4dq/8b98C3R92nH2I
9GxlhpfdtHlK0j5EiWUnIGh74+zfGzpLyra509q2pwJf4dEB4PTaDIL3Y8q7CeJL
LwFlKwjtO7P25qx7T2hAsKI9CH5zdnqcOB9N7bqjEhr2s4EoaZhu89Bx39laeOVu
5wPm1u+dONOG/moXoFKbkBNXygqLB0o9YV6hYtILgGdmQol+lmu+Ies3wlWTn58d
K18ovvqDp/RyFOM8mzMmL1S1Meu0AXj7eNP1hsDXpGH4izadBWCAGAN8y4z8cbdL
RbVsyVMNgFMBF02J/vcM7LmS3VoNTEOtHBp+RuJIKHAuZRmJHoR5j9P0bc2oXHoC
SWmAhKNPZIhEOf6URnqFLhY0MZhGFr1dOBk3wJAsPDQdD9lSN3KVk7v+dO5IGNjy
1rcDnCF0c8k93DTm8OqR4f6A1+R7scyDmEL70qSs2joR13Gq040hUM160BJdpxCy
Qqjo2+YG+gZRkAZCO6n/Vrnz8g9N8lD6x+PNgcqfRrE2O1MVr/c=
=A4RT
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on openpgp.org](https://keys.openpgp.org/vks/v1/by-fingerprint/E61773CD6E01040E2F1BD78CE7E2984B6289C93A)
</details>

ryanofsky commented on src/httprpc.cpp:
```diff
@@ -101,75 +100,37 @@ static bool RPCAuthorized(const std::string& strAuth, std::string& strAuthUserna
 }
 
-static bool HTTPReq_JSONRPC(const std::any& context, HTTPRequest* req)
+UniValue ExecuteHTTPRPC(const UniValue& valRequest, JSONRPCRequest& jreq, HTTPStatusCode& status)
```
re: https://github.com/bitcoin/bitcoin/pull/32297#discussion_r2819161479

Reason for calling the HTTP function instead of the JSONRPCExec is just that the HTTP code provides some additional functionality around raw JSONRPC code like user authorization and method whitelisting. I imagine in the future it could also provide more functionality too having different default wallets or different settings for different users.

It would be reasonable for this IPC inteface to just expose the JSON-RPC functionality not the HTTP functionality, but the IPC interface is used by `bitcoin-cli` so I think it makes sense to provide whatever `bitcoin-cli` needs to be fully functional.

ryanofsky reviewed (commented):
Thanks for the review!

pinheadmz reviewed (approved):
ACK 4565cff72c60f56d8ad4b46716fe8df847554824

Minimal diff since last review, fix conflict with change to mining interface.

Another thing i noticed is that `-rpcclienttimeout` is ignore when connecting by IPC. Not a deal breaker. If we add it back we should avoid libevent and I guess use CScheduler like https://github.com/bitcoin/bitcoin/pull/32862 ?

<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK 4565cff72c60f56d8ad4b46716fe8df847554824
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmmwKO4ACgkQ5+KYS2KJ
yToDmw/+LVtqNGxF9Tva6NLvhdNc23Yv1VUaYEB4M6jtm8CTj2qg/YFVJTCFfd1V
KEyz0eyJXEF+EALW1If0ujfGMfKMiM17DVxjeUA46BWCjPAxdNZj4qgIWTZz1CZ/
0ff+UYR+5eMPZHm0jJRfzRAC2Ch/ZVXDXN+ONraP0rTmEN8WyB/1gds5BgxsKe/x
Ks99Lm2TDnVzFWeILJgQKeNlJQ09DnpWqtZ4Pi23sfoUBSeucynRGUumT9gbz1ma
Sf7lDHle5fhw9yZJLGxPKuyu6uKoEcACs4vq143crQ1KiTKgnoiXXWZOvL+1rFev
3QdA0ewgVeOUWCzYWKmFJNHUGVxx0satbgvBIDgXWA9ydsMEQ3GBvguAY/41a29d
RrbRSnADchec8xD0YwmLXi0l1E0orcAXsDgcA95BQrfm5do2RtjVdk/OnybGws8N
cVmQ13N47jLjFn4zhtqZpF/w0iIO4hS+zfg5AF8FTRip+UJ/GpQ20rUdMaMB4+KX
lJsvB7MtLXg4H9FUnBMq+gFTJo3w0KQHDT18T8XNE0ExqFYd8Z+IgRw10QgsXfkQ
84o+RQYiVquRoDSUjGpYA4bObSfV/HFK0v46paepP8CuKNNTAYuSL4XjgZN1SnVG
8X8Utt4+R3Ykm/msihXVQR8+eHPPvJllO1uLT2y5cli6MB0kUIQ=
=JdCA
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on openpgp.org](https://keys.openpgp.org/vks/v1/by-fingerprint/E61773CD6E01040E2F1BD78CE7E2984B6289C93A)
</details>

enirox001 reviewed (commented):
Tested ACK 4565cff72c

Built on Fedora with ENABLE_IPC=ON. Ran `interface_ipc_cli.py`,
all tests pass.

Manually tested the following rpc commands

```
    bitcoin-node -regtest -ipcbind=unix -debug=ipc
    bitcoin-cli -regtest -ipcconnect=unix -getinfo
    bitcoin-cli -regtest -ipcconnect=unix getblocktemplate '{"rules": ["segwit"]}'
```

When node is stopped, -ipcconnect=auto correctly falls back to HTTP path.
